doc/source/data/examples/batch_inference_object_detection.ipynb

```diff
@@ -0,0 +1,350 @@
+    "# Batch Image Object Detection using Ray Data\n",
+    "\n",
+    "In this example, we will demostrate how to use Ray Data to do batch object detection with a PyTorch model on a large set of images.\n",
+    "\n",
+    "## Vanilla Pyorch exmaple\n",
+    "\n",
+    "Let's take a look at the [vanilla PyTorch example](https://pytorch.org/vision/stable/models.html#object-detection) first. \n",
+    "\n",
+    "In this example, we use a pre-trained PyTorch model to detect the objects on a single image. For simplicity, we skip the result visualization in this example.\n",
+    "\n",
+    "```python\n",
+    "from torchvision.io.image import read_image\n",
+    "from torchvision.models.detection import fasterrcnn_resnet50_fpn_v2, FasterRCNN_ResNet50_FPN_V2_Weights\n",
+    "\n",
+    "img = read_image(\"/path/to/your/image.jpg\")\n",
+    "\n",
+    "# Step 1: Initialize model with the best available weights\n",
+    "weights = FasterRCNN_ResNet50_FPN_V2_Weights.DEFAULT\n",
+    "model = fasterrcnn_resnet50_fpn_v2(weights=weights, box_score_thresh=0.9)\n",
+    "model.eval()\n",
+    "\n",
+    "# Step 2: Initialize the inference transforms\n",
+    "preprocess = weights.transforms()\n",
+    "\n",
+    "# Step 3: Apply inference preprocessing transforms\n",
+    "batch = [preprocess(img)]\n",
+    "\n",
+    "# Step 4: Use the model and visualize the prediction\n",
+    "prediction = model(batch)[0]\n",
+    "labels = [weights.meta[\"categories\"][i] for i in prediction[\"labels\"]]\n",
+    "print(labels)\n",
+    "```"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "## Parallelizing with Ray Data\n",
+    "\n",
+    "Then we want to apply this object detection model to a large set of images. We can use Ray Data to scale it.\n",
+    "\n",
+    "### Reading Image Data \n",
+    "\n",
+    "First, we use `ray.data.read_images` API to read image data from S3.  The directory structure of the dataset is`<s3_url/{label_id}/{*.JPEG}>`. So we use Partitioning utility to load in all the images for all labels. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ray\n",
+    "from ray.data.datasource.partitioning import Partitioning\n",
+    "\n",
+    "s3_uri = \"s3://anonymous@air-example-data-2/imagenette2/val/\"\n",
+    "\n",
+    "# The S3 directory structure is {s3_uri}/{class_id}/{*.JPEG}\n",
+    "partitioning = Partitioning(\"dir\", field_names=[\"class\"], base_dir=s3_uri)\n",
+    "\n",
+    "ds = ray.data.read_images(s3_uri, partitioning=partitioning, mode=\"RGB\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Understanding Batching\n",
+    "\n",
+    "To boost performance with hardware vectorization, we usually do inference in batches. In Ray Data, a batch is by default defined as a `Dict[str, np.ndarray]`. \n",
```

Comment:
I think let's split up the preprocessing and the inference steps?

Reply:
This example is a bit different. Both the preprocessing step and the inference steps share the `weights` argument. That's why I didn't separate them. But I can also pass the weights as an argument from the driver to each step, which may looks a lit bit weird. Which do you prefer?

Reply:
for preprocessing, we can just pass the transforms only from the driver to the UDF.

For the inference step, we would need to instantiate the weights inside the callable class. 

Last time I tried this, the weights themselves are not serializable.

Reply:
I think we can follow this: https://github.com/ray-project/ray/blob/fb1394e90a4a6bb11a0f315061a3230ab4ab4c2f/doc/source/ray-air/examples/pytorch_object_detection.ipynb? what do you think?

Reply:
sounds good.